### ISSUE TITLE
Store and retrieve state(s) from viewers in preferences

### ISSUE THREAD

**Issue by uliska** [wbsoft/frescobaldi#735]
This came up in #732.
It's not clear yet whether we need a "Viewers" Preferences page, but in any case I think some stuff should be saved/restored using the QSettings interface.
Among them should be the visibility of a viewer's toolbar and the "synchronize with cursor" option. As the latter is already saved we should look into that for a start.

**Comment by uliska** [comment]
This has to take into account that there are more than one viewers to maintain

**Comment by uliska** [comment]
@wbsoft I'm trying to read the code in `qpopplerview` but don't seem to get anwhere.
Can you tell me where to look for properties that I can retrieve and set to store/restore the current position of a manuscript/music preview in the settings?

**Comment by PeterBjuhr** [comment]
I think what we are looking for is `view.position` and `view.setPosition`. Note also that in `AbstractPopplerView` we have the variable `self._positions`.

**Comment by uliska** [comment]
I've now used `view.position` to save/restore the position of the *currently active* manuscript.

This is only half the way, but I think we'll have to reconsider the maintenance of different manuscripts/docs in a more fundamental way (#737), so I would rather not bother going further here *now*.

**Comment by PeterBjuhr** [comment]
You haven't pushed this change, or have I missed something?

**Comment by uliska** [comment]
Sorry. Now it's there. 198297c4

**Comment by PeterBjuhr** [comment]
Thanks!

**Comment by PeterBjuhr** [comment]
For some reason I haven't figured out yet the manuscript can't be correctly re-positioned when the app is reopened. But it works when closing and opening a session without closing Frescobaldi.

**Comment by uliska** [comment]
Are you really sure about this? On my computer the *current* manuscript opens at the last position (kinetically scrolling there if it is active). Either there is something with your seting set-up or the thing is somehow system-dependent.

**Comment by uliska** [comment]
Of course (as said) he behaviour is *only* correct for the current document.

**Comment by PeterBjuhr** [comment]
Does it work also without the kinetic scrolling?

When I tested the scrolling always stopped at page 7. Without the scrolling it didn't work at all.

As you point out it can be something in my system that prevents this from working correctly.

(By the way, I think it is preferable without using the scrolling...)

**Comment by uliska** [comment]
Oh, you're right. Same behaviour here, only that Kinetic scrolling stops at page 3 already. Very strange indeed. I have no idea what could be the reason for this.

And I agree that it would be preferrable to disable kinetic scrolling when opening a document.

**Comment by uliska** [comment]
It seems `kineticMove` applies some energy that doesn't completely match our case when starting from a new app instance. I've applied a `print`debug in the `test-kinetic-move` branch, and this shows an interesting behaviour when repeatedly restarting Frescobaldi with a long manuscript open.

From `kineticscrolling.KineticScrollArea.kineticMove```:

```
# Since this function is called for exact moves (not free scrolling)
# limit the kinetic time to 2 seconds, which means 100 ticks, 5050 pixels.
```

If the following lines are commented out the scrolling *nearly* works, even when it needs an inacceptably long time to reach the destination. However, it's still a few cm away from the actual position.

I think we are facing a design flaw in the Kinetic Scroll module that hasn't appeared yet because the current music view doesn't try to restore the position.

I think we should find out what is happening and make it work after suppressing the kinetic scrolling completely. When opening the app or a session there is absolutely no sense in having kinetic scrolling.

**Comment by PeterBjuhr** [comment]
So the scrolling takes longer time when Frescobaldi restarts? This still doesn't explain why it doesn't work without the scrolling, i. e. with ` overrideKinetic=True` in `setPosition`.

**Comment by uliska** [comment]
Absolutely not. I'm right now starting to look into it.
Yes, the scrolling can take ages with that change, and it stops *somewhere* around the intended point but not exactly.

I think the following is necessary:

* Suppress kinetic scrolling upon session change (not only on startup)
* Determine why it doesn't work (only on application startup) without kinetic scrolling
* Fix the latter

**Comment by uliska** [comment]
I've applied some "print debugging" in the `debug-kinetic-scroll` branch (pushed that so you can compare with your own results. There are interesting results and differences between output given on startup and session switching.

This is the output of a successful session switch:

```
session changed to: Default
Position from session: (0, 0, 0)
new position as point: PyQt4.QtCore.QPoint(4, 4)
Point going into fastCenter: PyQt4.QtCore.QPoint(4, 4)
Calculated diff to viewport: PyQt4.QtCore.QPoint(-495, -210)
Position from session: (6, 0.49949545913218973, 0.38659058487874465)
new position as point: PyQt4.QtCore.QPoint(499, 9006)
Point going into fastCenter: PyQt4.QtCore.QPoint(499, 9006)
Calculated diff to viewport: PyQt4.QtCore.QPoint(0, 8792)
Point going into fastCenter: PyQt4.QtCore.QPoint(499, 9006)
Calculated diff to viewport: PyQt4.QtCore.QPoint()
```

* (`Default` is the explicit name of my session switched to)
* I assume the `(4,4)` refers to the margin around the "paper"
* `(-495, -210)` seems to refer to the actual viewport. These numbers change when I resize the viewer.
* The `(6, 0. ...)` is probably the stored position halfway through page 7
* `(499, 9006)` is the correct target position, followed by the calculated diff to the viewport

Interestingly `View.setPosition()` is called three times.
Research turns out the following:

* First execution (with position `(0, 0, 0)` happens implicitly when `loadManuscripts()` calls `self.updateDocument()`. I have the impression this is unnecessary and should rather be suppressed some way, but it doesn't seem to hurt either.
* Second execution (with position `(6, 0.499, 0.38)` is the one called from within `slotSessionChanged()`
* The third one, passing the same Point into the function, is a mystery to me.

This is the output of restarting the application:

```
session changed to: Default
Position from session: (0, 0, 0)
new position as point: PyQt4.QtCore.QPoint(4, 4)
Point going into fastCenter: PyQt4.QtCore.QPoint(4, 4)
Calculated diff to viewport: PyQt4.QtCore.QPoint(-495, -210)
Position from session: (6, 0.49949545913218973, 0.38659058487874465)
new position as point: PyQt4.QtCore.QPoint(499, 9000)
Point going into fastCenter: PyQt4.QtCore.QPoint(499, 9000)
Calculated diff to viewport: PyQt4.QtCore.QPoint(0, 8786)
Point going into fastCenter: PyQt4.QtCore.QPoint()
Calculated diff to viewport: PyQt4.QtCore.QPoint(-499, -9000)
```

There is one single difference here, and that seems the culprit: The last time there is no point passed to `setPosition`, and that seems to make the position be kind of reverted back to the top of the file.

So it seems we have to identify from where `setPosition()` is called the third time and find out why it behaves differently when switching sessions or when starting the application.

Any ideas @PeterBjuhr ?

**Comment by PeterBjuhr** [comment]
Yes, I had a similar mysterious third calling yesterday when I tested. It turned out the answer was pretty obvious. I removed the music viewer and the mystery was solved.

I think it is a good idea to remove the music viewer while testing to not let that interfere. I hope that this also will be helpful for you!

**Comment by uliska** [comment]
What do you mean with "removing", is closing enough? This doesn't change anything.

**Comment by uliska** [comment]
I think I have a new hint. The first and the third invocations always go into the `fastCenter` method while the middle one (our own) depends on the kinetic scrolling settings: it'll go to either `fastCenter` or `center`.

It seems there's a call *without* a point going into `fastCenter` at the end of the process in any case. I suspect that this is simply overridden by the timers or whatever drives the kinetic movement. So this would mean: also with kinetic scrolling enabled the position is reset to the top of the file but the kinetic scrolling simply continues afterwards. Maybe this could even explain the small difference in the target point when I made the scrolling go to the end (i.e. removing the explicit limitation).

**Comment by uliska** [comment]
So the task is to identify where this third call is coming from.

**Comment by PeterBjuhr** [comment]
>What do you mean with "removing", is closing enough? 

Yes closing.

>This doesn't change anything.

Ok, It does call `setPosition()` on opening so I'd expect that to make a difference...

>The first and the third invocations always go into the fastCenter method while the middle one (our own) depends on the kinetic scrolling settings: it'll go to either fastCenter or center.

I think the ones that goes into fastCenter comes from `openDocument` in the widget. There `setPosition()` is called with `overrideKinetic=True`.

**Comment by uliska** [comment]
It is the *first* call that comes from `openDocument`. 
But I just found out that the third call is triggered in
`qpopplerview.View._resizeTimeout`. I'm experimenting here, and it seems I can get the stuff to work from here. Just have to be careful about side effects.

**Comment by uliska** [comment]
I *think* I've fixed the issue in bfa65fa2. Phew!

It turned out that the third call was triggered in the viewer's resize event. This happens *after* setting the position, and it only happens when the application is created but not when sessions are changed.

However, we still have only the *current* document's position saved to and restored from the session.

**Comment by uliska** [comment]
I'm not clear where the position of the non-current documents are stored, so I don't know how to retrieve them in `AbstractPopplerView.saveSession()` yet.

**Comment by uliska** [comment]
@PeterBjuhr what do you think? #737 will presumably significantly affect how these informations are stored and managed. So maybe we could settle with the current state (restoring the position of the *current* document only), open a new issue for the next milestone, depending on #737, and close the current one?

**Comment by PeterBjuhr** [comment]
>It turned out that the third call was triggered in the viewer's resize event. This happens after setting the position, and it only happens when the application is created but not when sessions are changed.

The resize event and the `resizeTimeout()` does get triggered also when sessions are changed (and also after your commit). But it works so you do seem to somehow prevent the reset of position. Did you figure out why the position gets destroyed in the resize process?

**Comment by uliska** [comment]
I have to look into the first part of your comment.
But the thing that mainly blocked the thing to work was:

* First the position was set to "0" (in openDocument())
* Then the resize timer was triggered
* Then the position was set correctly
* When the timer was executed the position was reset to "0"

*With* kineticScrolling the timer *also* reset the position, but the kinetic scrolling was simply continued so one didn't notice the problem.

**Comment by PeterBjuhr** [comment]
It should be possible to resize without the viewer loosing the position. So a call to resize in itself isn't a problem. And as I mentioned above when I tested just now the resize function does get called without anything bad happening. So I think the question remains why does this error occur when restarting 
Frescobaldi with a session!??

**Comment by uliska** [comment]
The *trigger* of the problem was that the last call (from the timer)
used the (0, 0, 0) position.
Maybe the real solution was to suppress the (0, 0, 0) from openDocument.
I can imagine that it was that value that the timer picked up.

**Comment by uliska** [comment]
The thing is that of course the resizeEvent is triggered, but now there's the check for `_inSessionChange` that prevents `_resizeTimer` to be triggered.

**Comment by PeterBjuhr** [comment]
As I mentioned above I also checked if `resizeTimeout()` was in fact reached.

**Comment by uliska** [comment]
Hm, I can't confirm that.
Adding `print "reached"` inside `View._resizeTimeout()` doesn't print anything, only when really resizing the viewer.
?

**Comment by PeterBjuhr** [comment]
I get two "reached" when restarting Frescobaldi (with your commit in place) and two "reached" when opening session.

**Comment by PeterBjuhr** [comment]
>Maybe the real solution was to suppress the (0, 0, 0) from openDocument.
I can imagine that it was that value that the timer picked up.

That sounds somehow plausible. But in fact testing tells me that even if the position isn't set at all in `openDocument` the error still occurs.

**Comment by uliska** [comment]
To make sure we're talking about the same stuff I have added a debug commit to a new `debug-resize-event` branch. This adds debug prints in three positions:

* entering `resizeEvent()`
* entering `_resizeTimeout()`
* the line where `_resizeTimer` is triggered

On my system `resizeEvent` is reached several times, but `_resizeTimer` is *only* triggered when actually resizing the viewer.

If you see different results we seem to have some system dependent behaviour with regard to PyQt's signal handling.

**Comment by uliska** [comment]
Oops, forgot an important part. I get four resize events on application start, but not a single one when switching sessions.

**Comment by PeterBjuhr** [comment]
This is the output from starting Frescobaldi:

````
reached resizeEvent()
self._inSessionChange: False
reached resizeEvent()
self._inSessionChange: False
reached resizeEvent()
self._inSessionChange: False
trigger _resizeTimer
reached _resizeTimeout()
self._inSessionChange: False
reached resizeEvent()
self._inSessionChange: True
reached _resizeTimeout()
self._inSessionChange: False
````

**Comment by PeterBjuhr** [comment]
This is when (re)opening session:

````
reached resizeEvent()
self._inSessionChange: False
reached resizeEvent()
self._inSessionChange: False
reached resizeEvent()
self._inSessionChange: True
reached resizeEvent()
self._inSessionChange: False
trigger _resizeTimer
reached _resizeTimeout()
self._inSessionChange: False
reached _resizeTimeout()
self._inSessionChange: False
````

**Comment by uliska** [comment]
OMG. Here it is

```
self._inSessionChange: False
reached resizeEvent()
self._inSessionChange: False
reached resizeEvent()
self._inSessionChange: False
reached resizeEvent()
self._inSessionChange: True
```

What are your settings for sessions during startup and inside the concrete sessions?

**Comment by PeterBjuhr** [comment]
Always save and start with last used session.

**Comment by uliska** [comment]
Ah, now I see a difference: I didn't have a document in the music viewer, only in the manuscript view. When I have a document in the music viewer I get the same output as you.

So obviously the confusing events happen in the music viewer. The `View` class is used both in the new viewer and the old.

**Comment by uliska** [comment]
With the music view completely closed I get

```
reached resizeEvent()
self._inSessionChange: False
reached resizeEvent()
self._inSessionChange: True
```
on startup

**Comment by PeterBjuhr** [comment]
Yes, I forgot to close the music viewer like I did yesterday. Sorry about that! But I think the point still holds, we don't really know why the resize event can't handle the situation.

And we don't really know what happens differently on startup, or do we?

**Comment by PeterBjuhr** [comment]
One more thing to report. I'm not sure it's totally unrelated to this issue. If it is, I can open a new issue.

When opening a session the previous position is set, but when I hide the toolbar the position is suddenly lost.

**Comment by uliska** [comment]
This looks only loosely related but actually was introduced in bfa65fa2

**Comment by PeterBjuhr** [comment]
I think I can confirm it is related - similarly an actual resize also resets the position, contrary to the normal behaviour. 

I think we can confirm that just preventing the resize to get triggered isn't really solving the problem. Something makes the set position very unstable and we have to figure out the problem. Or find a totally different approach.

**Comment by uliska** [comment]
Yes, you're right. But I for my part have to defer that to tomorrow - my mental capacity for today is "over".

**Comment by PeterBjuhr** [comment]
>This looks only loosely related but actually was introduced in bfa65fa

Interesting! Of course it wasn't possible to set the position at all previously. But I just tested and the problem also occurs when switching session (not on app startup). Do you mean that this was introduced in that commit?

**Comment by uliska** [comment]
I just checked out the previous and that commit, and it seems this commit introduced the issue with hiding/showing the toolbar.

Well, probably that's quite simple: hiding the toolbar triggers a resize event but doesn't set `_inSessionChange`. At least that's a start understanding why this commit changed *something*.

**Comment by PeterBjuhr** [comment]
Yes, I didn't state that clearly - it must be because the hiding of the toolbar triggers a resize we get this problem.

Your commit is postponing the reszing problem, but does it also introduce it in the case of switching sessions which worked before!?

**Comment by uliska** [comment]
No idea. Sorry, will shut down the computer now ...

**Comment by uliska** [comment]
I can confirm that checking out

* 409d5025

**Comment by uliska** [comment]
I can confirm the following behaviour with commits checked out:

* 409d5025  
-- Wrong position on startup with kinetic scrolling on  
-- no positioning on startup without kinetic scrolling.  
-- correct behaviour on session change  
-- correct behaviour when toggling the toolbar
* bfa65fa2 (after the fix)  
-- Correct behaviour on application startup, regardless of kinetic settings  
-- Correct behaviour on session change  
-- toggling toolbar causes fastScroll to top of score.

So obviously what I did there caused the bug with the toggling.

**Comment by PeterBjuhr** [comment]
As suggested above, what you did was probably to move the problem from the resize on startup till the resize when toggling the toolbar.

Are you reverting the commit, or do we progress from here? 

I want to try a approach were we save and restore the positions from `self._positions` (to the session) and not directly from the viewer. If it's successful we should be able to retain all positions and not just from the current document.

**Comment by uliska** [comment]
I've reverted that commit (including some merge conflicts), and the `debug-startug-positioning` branch should now give a "clean" start for a new approach. Maybe you go ahead and give your idea a try (which might also allow to store *all* positions). I'll do something different then this morning (in order not to interfere) and just watch out for any comments or questions you have.

Just one hint: When I first touched the `_positions` WeakRefDictionary I had the impression this isn't really reliable - as when I wanted to access it it was actually empty.

**Comment by PeterBjuhr** [comment]
It was just an idea from a brief encounter in the code. I haven't tested yet if it will do what I hope it can!

**Comment by PeterBjuhr** [comment]
The approach is successful in saving positions from all manuscripts. But the problem on startup persists. If I had an idea how to solve that too, it's (temporarily?) lost.

>Just one hint: When I first touched the _positions WeakRefDictionary I had the impression this isn't really reliable - as when I wanted to access it it was actually empty.

It doesn't get set until the view is cleared with a new document, therefore I also stored the current document and its position.

**Comment by PeterBjuhr** [comment]
Following through @uliska excellent debugging (of the issue at startup) that led us to the `resizeEvent` I now more or less fully grasp the problem. In short, the first time a resizing is fully carried through the current position is not used but instead `QPoint(0, 0)` which lead any previously set position to get lost.

I think it will possible to find a workaround and somehow trigger the resize event before the position is set. But I'm not sure of this is a desirable behaviour in the qpopplerviewer.

**Comment by PeterBjuhr** [comment]
@uliska can you confirm that the functionality now is ok?

**Comment by PeterBjuhr** [comment]
Before closing this there are two more things I think we shall consider:

1. The visual relation/proportion between the music view and the manuscript view is not stable. I often have to manually restore this. Anyone else that is experiencing this?

2. It would be great if the zoom factor could be saved. The first thing I have to do is to restore this.

**Comment by fedelibre** [comment]
I don't understand the first point.

I agree on the second, even if I wouln't put it in this initial milestone (unless you already know how to implement it). As a workaround, you can set the viewers width to fit all the space available and stretch the panel when you need to zoom (this is saved to the session).

**Comment by uliska** [comment]
Currently I have an issue with empty documents that cause a crash (because somehow upon saving or restoring they are "given" a filename of `/`. I'm right now trying to understand what's wrong.

One thing is that the current code doesn't discern between new and empty sessions because it only checks the presence of a current manuscript.
I'll be looking into this right now.

**Comment by uliska** [comment]
I've fixed one more thing in that context: When saving an empty session it still iterated over `documents`, which caused a crash.

See a24e472e, but it doesn't help the loading issue.

**Comment by PeterBjuhr** [comment]
>Currently I have an issue with empty documents that cause a crash (because somehow upon saving or restoring they are "given" a filename of /. I'm right now trying to understand what's wrong.

I haven't experiencing anything like this. Does it happen with all documents?

>One thing is that the current code doesn't discern between new and empty sessions because it only checks the presence of a current manuscript.

See 77eb9caecd5ad47399bf64296c6e0cbb707021b5. Or do you mean a session without an active document (in the editor)? I didn't consider that...

**Comment by PeterBjuhr** [comment]
>I don't understand the first point.

Sorry, let me be more precise. I aim at a separation between the music and manuscript viewer where they get approximately half of the horizontal space each. But often the manuscript viewer is considerably smaller when I start Frescobaldi.

>I agree on the second, even if I wouln't put it in this initial milestone (unless you already know how to implement it). As a workaround, you can set the viewers width to fit all the space available and stretch the panel when you need to zoom (this is saved to the session).

I don't think the implementation is hard, but maybe we should wait anyway so the viewers are more in sync regarding their functionality.

I'll take a look at the workaround, thanks!

**Comment by uliska** [comment]
A session that doesn't have an active document in the editor is consequently empty. I think my commit fixed *that* issue.

With 6b3bb810 I try to make the new session check more explicit, and I think it works now. I don't get the "empty" items in the document chooser anymore, and I don't get crashes.
However, I have the impression that not all manuscripts are reliably saved. My impression is that it has to do with the weakrefs that are not guaranteed to be present. Sometimes after reopening a session only one or two files are restored, and maybe it's that only files that have already been at the front are saved. I'm not really clear about it.
But I'm rather sure that for a reliable save it is necessary to iterate over the list of files that is currently stored in the document chooser.

**Comment by uliska** [comment]
0cbb1f28 implements this, and I am quite convinced saving and restoring *all* files and handling empty sessions now works.

I haven't checked what happens with the positions, though. (will do now).

**Comment by PeterBjuhr** [comment]
Yes, I think it's a good idea (but haven't experienced any problems myself)

**Comment by PeterBjuhr** [comment]
>I haven't checked what happens with the positions, though. (will do now).

I did a quick test and it seems it still works!

**Comment by uliska** [comment]
What did you change about the position storage?
Now it seems that positions are correctly loaded also on application startup!

**Comment by PeterBjuhr** [comment]
The problem wasn't the storage of the positions, the problem was the resize event as you rightly debugged earlier. The position was correctly set but it was immediately washed away by the initial resize. So the view has to know that a position has been set to avoid the default resize reset, see f29d74bd26a21124a4f6d1164967a93d8752b7bf (if nothing happens I indend to make this into an own pull request). I tried to find a temporary workaround (for the manuscript branch) but all I came up with was 02f3da64485083423d1498b7f7dd174b67caf28a.

**Comment by PeterBjuhr** [comment]
I just realized that I've forgot to tell those who are testing out the manuscript branch that sessions that were previously (before the weekend) saved with manuscripts will end up with an error. We are now also storing the positions so the system have changed somewhat.

**Comment by uliska** [comment]
I think 25b63650 and the previous commit now make it work properly (apart from still using the hack to suppress the resize event).

Maybe we can now finally close this???
I suppose you are aware that we are struggling with some functionality that the current Music View doesn't have. So when we will integrate the viewers we'll automatically improve the Music View with this :-)

**Comment by uliska** [comment]
Nothing has popped up in the last week so I consider this finished.

**Comment by uliska** [comment]
OMG. It *did* seem to work, but actually the restored positions are wrong, at least with my latest testing. No idea what went wrong here ...

**Comment by uliska** [comment]
But - again - only on application startup.

**Comment by uliska** [comment]
OK, happened in c474727f.
Will have to look into it later this evening.

**Comment by uliska** [comment]
The reason is that the change in this commit triggers a `setVisible` and presumably a resize event later in the process. However, wrong startup position always happens when the program is restarted *after changing* the visibility of the toolbar. Otherwise it seems OK.

**Comment by uliska** [comment]
I'm not so sure anymore that it's the resize event this time. What I'm stumbling over is the fact that the positions seem fixed. That is when I change the toolbar visibility I'm getting a different position upon restart. But when I switch visibility back again I'm getting the same position as before. So positions somehow toggle along with the visibility.

My suspicion is (although this is only an impression and not based on any insight from the code) that somehow switching visibility of the toolbar at that later point messes up the position in relation to the actual widget size.

**Comment by PeterBjuhr** [comment]
What you have this time is a small diff in position, right?

**Comment by uliska** [comment]
Dependibg on the scalings the diff can be quite significant (up to 1/3 page). But it is consistent.
But I'm sure it us related to the difference in "position" before and after showing/hiding the toolbar. Somehow this difference is not respected or accounted twice.
-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by PeterBjuhr** [comment]
You changed the session handling don't you think that has made a difference?

**Comment by uliska** [comment]
We could re-check, but I tried on a separate branch to cherry-pick the individual commits in different order. With the session handling commit it still works, only when adding the toolbar changes the bad behaviour pops up.

**Comment by uliska** [comment]
I *did* re-check, and yes, with 59c0a9ba (session handling) applied it still works correctly, but with 8485d215 (toolbar) it is wrong.

The offset can be substantial, even for several pages. I have the impression that it is somehow related to the length of the whole file, i.e. the relation of the toolbar height to the View height.

**Comment by PeterBjuhr** [comment]
Thanks, it helps debugging!

**Comment by uliska** [comment]
I've done some more debugging by printing positions before and after toggling the toolbar:

```bash
Debug slotShowToolbar.
Do show toolbar: True
Value before toggling: (23, 0.49837662337662336, 0.34288990825688076)
Value after toggling: (23, 0.49837662337662336, 0.3165137614678899)

Debug slotShowToolbar.
Do show toolbar: False
Value before toggling: (23, 0.49837662337662336, 0.3165137614678899)
Value after toggling: (23, 0.49837662337662336, 0.3165137614678899)

Debug slotShowToolbar.
Do show toolbar: True
Value before toggling: (23, 0.49837662337662336, 0.34288990825688076)
Value after toggling: (23, 0.49837662337662336, 0.3165137614678899)
```

Initially the viewer is at a random position on page 23 (actually 24 of course).
Switching the toolbar on makes the center of the view be at a slightly smaller ratio of the page => there is a little less space visible of the page, so the center move up by half of the toolbar's height. So *this* output is understandable.

What looks strange to me is the second set, switching the toolbar off again. It starts with the right value (the same we ended before), but it reports *the same* position after the toggle. I would have expected the original value we started off with, as the visible appearance is reverted to the original state.

The third one is somewhat strange too. While the numbers are the same as in the first run (which is expected) I don't know how we got from the end of the second run (the `0.3165...`) back to the start of the third one (the `0.34..`)

I have the strong impression that this inconsistency may be the cause of the issue.

**Comment by uliska** [comment]
Some fruitless attempts later I'm tempted to leave that issue alone and live with the widget being created upon startup. But of course that's not a real alternative, isn't it?

I noticed that when the widget is created (given that the panel is opened) there is no current session available yet.

**Comment by PeterBjuhr** [comment]
I found out that the temporary hack that prevents `resizeEvent` from deleting the set position is now responsible for the new issue - when removing the hack everything works as expected but of course the saved position is gone again. (I don't know exactly why this issue was avoided previously.)

Anyway, as the problem is in qpopplerview or even in Qt itself and not in the viewer I think this can be closed.

**Comment by uliska** [comment]
It has to be noted that it only affects the active document. My test indicate that if you have several manuscripts open the other ones properly restore their position.

I see two questions regarding closing the issue:

1. Is it OK to keep the current behaviour? Positions are (from the perceptions of the user) "saved" for all documents except the active one, and the position of the active one is somewhat random (instead of the behaviour of the current musicview where documents always open at the top). Or should we rather suppress the functionality completely (so the new viewers behave just like the existing one)? In the latter case I'd strongly vote for *suppressing* the behaviour and not reverting/purging it completely.
2. I would say this should be opened as a new issue again. I still think we should get this stuff to work some day (even if it should be a Qt bug it might be solvable after switching to PyQt5), but not have it block our initial release.

**Comment by PeterBjuhr** [comment]
I'm sorry, I was unclear! I didn't close this because I think the problem is solved (in fact I'm thinking of the problem very actively!). But I do think it's good to get a fresh start so I create a new issue!

**Comment by uliska** [comment]
My impression was rather you considered the problem unsolvable or "not our business". 

However,  I can very well imagine leaving that open for a future solution. But the initial behaviour should be consistent in itself.
